### PR TITLE
TypeScript support for CLI function-generator

### PR DESCRIPTION
**Pull request by realStandal**
This is part of #523 (thank you @kimadeline for doing `component` and your PR :) )

The issue looks like it was at a bit of a stand-still, something about `cli` being split into its own repo as well. Regardless, it was nice learning from doing it.

- Added use of `yargsDefaults` to provide `--javascript`, `--typescript`, and `--force` options.
  - Includes `--js` and `--ts` aliases. 
  - `--javascript` is the default as per `yargsDefaults` configuration.
- Template converted to TypeScript
  - Template is transpiled to JavaScript when the command is run with the option `--javascript`.
  - Template will lose type annotations during transpilation.
- Altered existing tests to take into consideration new options.
- Added tests that cover `--js` and `--ts` options.
- Refactored [`helpers.test.js`](https://github.com/redwoodjs/redwood/blob/main/packages/cli/src/commands/generate/__tests__/helpers.test.js) to use updated `function.ts.template`.
- Ensured `destroy` command works on generated `.js` and `.ts` files (can this be tested?)

In terms of what this **doesn't** include, it's the same as what's described in PR #632 minus any `web` specifics. That includes mainly non-command-specific implementations.

I'm fairly certain this is my first contribution to a project that isn't my own; hopefully, all ducks are in a row :)
Let me know if anything isn't right, and thank you for providing such straight-forward direction on accomplishing this.

### CODE REVIEW

File: packages/cli/src/commands/generate/function/function.js [hunk 1]
```diff
@@ -3,48 +3,58 @@ import path from 'path'
...
     generator: 'function',
-    templatePath: 'function.js.template',
+    templatePath: 'function.ts.template',
     templateVars: { ...rest },
-    outputPath: path.join(getPaths().api.functions, `${functionName}.js`),
+    outputPath: path.join(getPaths().api.functions, `${functionName}${isJavascript ? '.js' : '.ts'}`),
   })
 
-  return { [file[0]]: file[1] }
+  // If --javascript=true convert TS->JS before returning
```
Comment (Tobbe) [inline packages/cli/src/commands/generate/function/function.js:30]: I think this comment is redundant. It's clear from the code what's going on here

File: packages/cli/src/commands/generate/function/function.js [hunk 2]
```diff
@@ -3,48 +3,58 @@ import path from 'path'
...
+import { yargsDefaults } from '../../generate'
 import {
   templateForComponentFile,
   createYargsForComponentGeneration,
 } from '../helpers'
 
-export const files = async ({ name, ...rest }) => {
+export const files = ({ name, ...rest }) => {
+  // Taken from ../component; should be updated to take from the project's configuration
+  const isJavascript = rest.javascript && !rest.typescript
```
Comment (Tobbe) [inline packages/cli/src/commands/generate/function/function.js:16]: I'd add `const extension = isJavascript ? '.js' : '.ts'` and then use that on line 22 and 27

**Comment by Tobbe** [comment]
@realStandal Thanks for taking this on! Much appreciated :)

**Comment by realStandal** [comment]
Not a problem, In regards to what you've commented on, I just make the changes push to my repo and Git(Hub) takes care of updating this PR, right?

**Comment by Tobbe** [comment]
@realStandal Yeah, pretty much. There are a couple of way of doing it, but it mostly comes down to personal preference. Some people create a new commit for every change and push that. Others like to amend the existing commit if it's just small bugfixes or tweaks and then force-pushing that. In the end we'll likely squash all your commits if you have several anyway, so do what you feel most comfortable with 🙂

**Comment by Tobbe** [comment]
And also, always feel free to argue my suggestions if you don't agree with my code review 👍

**Comment by realStandal** [comment]
Sounds good! There are those two changes.

Thanks for letting me know! I do agree with those two, of course lol.

**Comment by realStandal** [comment]
Also just wanted to make the note that documentation, specifically [here](https://redwoodjs.com/docs/cli-commands#function), will need to get updated.

**Comment by Tobbe** [comment]
Thanks for highlighting the doc-changes needed. Could you please create an issue for it? Or, even better, a PR?` (both in the redwoodjs.com repo)

Could you please also fix the lint errors in the code here? Thanks!

**Comment by realStandal** [comment]
Sure, I ran it and it complained about wanting it like this; I originally had it how GitHub's wanting it, maybe I made the line longer. And I will do the PR, didn't know if y'all were waiting and keeping track between releases.

**Comment by Tobbe** [comment]
> And I will do the PR, didn't know if y'all were waiting and keeping track between releases.

Thank you! What will happen is that someone will approve the PR, but it won't get merged until this code has made it into a proper RW release 🙂

File: packages/cli/src/commands/generate/function/templates/function.ts.template [hunk 3]
```diff
@@ -0,0 +1,10 @@
+import { APIGatewayEvent, Context } from 'aws-lambda'
```
Comment (Tobbe) [inline packages/cli/src/commands/generate/function/templates/function.ts.template:1]:
```suggestion
import type { APIGatewayEvent, Context } from 'aws-lambda'
```

Comment (realStandal) [reply packages/cli/src/commands/generate/function/templates/function.ts.template]: This'll be cool cause I did not know about this functionality and it'll be provided for all TS-functions.

File: packages/cli/src/commands/generate/function/templates/function.ts.template [hunk 4]
```diff
@@ -0,0 +1,10 @@
...
+
+export const handler = async (event: APIGatewayEvent, context: Context) => {
```
Comment (peterp) [inline packages/cli/src/commands/generate/function/templates/function.ts.template:3]: I would love to add some syntax to help our none-typescript users out.

```suggestion
/**
 * The handler function is your code that processes http request events.
 * You can use return and throw to send a response or error, respectively.
 * 
 * @param { import("'aws-lambda'").APIGatewayEvent } event - an object which contains information from the invoker.
 * @param { import("'aws-lambda'"). Context } context - contains information about the invocation,
 * function, and execution environment.
 */
export const handler = async (event: APIGatewayEvent, context: Context) => {
```

Also, does AWS lambda has a return type?

Comment (peterp) [reply packages/cli/src/commands/generate/function/templates/function.ts.template:3]: I'm not 100% sure that that syntax will work. If it doesn't then we'll have to use @typedef.

Comment (Tobbe) [reply packages/cli/src/commands/generate/function/templates/function.ts.template:3]: @peterp Why the double quotes in the import?

This is how you'd import with typedef
```
/**
 * @typedef {import('./Foo').default} Bar
 */
 ```

Comment (peterp) [reply packages/cli/src/commands/generate/function/templates/function.ts.template:3]: Ah, thanks for catching that... turns out textareas aren't great code editors.

Comment (realStandal) [reply packages/cli/src/commands/generate/function/templates/function.ts.template:3]: Didn't get a notification for this sorry. [Netlify's documentation](https://docs.netlify.com/functions/build-with-javascript/#synchronous-function-format) is where I got direction for the parameter types. [This article](https://serverlessfirst.com/fff-aws-lambda-type-definitions/) provides a return type from the same package, haven't tested that as I just found it. I've never used lambda so I don't know what it's expecting as a result.

**Comment by thedavidprice** [comment]
Hi all! Checking in about status here as we'd like to get this into the next release. What's the next step needed?

**Comment by Tobbe** [comment]
I'm happy with this, just need to update the code comment as per @peterp's suggestion.

**Comment by realStandal** [comment]
I'm horribly sorry, I swear I came back to check on this and still didn't notice y'all's conversation or the suggestion.

**Comment by realStandal** [comment]
Do the changes to main need to come in on top?

**Comment by Tobbe** [comment]
Thanks for updating the comments! Did you try it out? Mind sharing a screenshot of how VSCode displays that when writing a Function?

**Comment by realStandal** [comment]
Sure, here are both the recommendations y'all had made:

Screenshot from 2021-03-18 13-17-53

Screenshot from 2021-03-18 13-19-55

**Comment by Tobbe** [comment]
I wonder why it isn't displaying the types. @peterp do you know?
Like here 
image
I want it to show the type of `event`, not just its name. Maybe VSCode doesn't do that.

**Comment by realStandal** [comment]
> I wonder why it isn't displaying the types. @peterp do you know?
> Like here 
> image
> I want it to show the type of `event`, not just its name. Maybe VSCode doesn't do that. 
> 

@Tobbe Did you notice above the commented part by chance? Theyre not links to open the source, but it includes the type on the param above; where it gives you the hint at what the functions signature looks like.

**Comment by Tobbe** [comment]
> > I wonder why it isn't displaying the types. @peterp do you know?
> > Like here
> > image
> > I want it to show the type of `event`, not just its name. Maybe VSCode doesn't do that.
> 
> @Tobbe Did you notice above the commented part by chance? Theyre not links to open the source, but it includes the type on the param above; where it gives you the hint at what the functions signature looks like.

@realStandal Ohh, you're right. Missed how it looks for the one with @param. That's much better than what you got for @typedef, where it only said `any`. But still, what's the point of the whole `import { ... }` stuff if it's not pulling any info out from there?

I'm sorry. I feel like I'm just being difficult here. You've done great work Ryan, I'll go read up on JSDoc and the import stuff on my own. Thanks again for doing this work!

**Review by Tobbe — APPROVED** [review]

**Comment by realStandal** [comment]
@Tobbe You're perfectly fine, I've never understood `@params` proper use anyway, nice excuse to learn more. From what I see, that's just how VSCode handles showing the message. Here is a snippet from the [JSDoc on @param](https://jsdoc.app/tags-param.html).

```JavaScript
/**
 * @param {string} somebody
 */
function sayHello(somebody) {
  alert('Hello ' + somebody)
}
```

Which results in

image

I agree, having it below makes more sense; took me awhile to figure out the signature above gave me the types. Adding a `{@link ...}` to references those types could be an option-at least that way it's right next to where a user would be looking; I'm not sure where those references are, though.

**Comment by peterp** [comment]
@Tobbe I did it like this:
image

File: packages/cli/src/commands/generate/function/templates/function.ts.template [hunk 5]
```diff
@@ -0,0 +1,18 @@
...
+ * function, and execution environment.
+ */
+export const handler = async (event: APIGatewayEvent, context: Context) => {
+  return {
+    statusCode: 200,
+    body: JSON.stringify({
+      data: '${name} function',
+    }),
+  }
+}
```
Comment (peterp) [inline packages/cli/src/commands/generate/function/templates/function.ts.template:18]:
```suggestion
/** @typedef { import('aws-lambda').APIGatewayEvent } APIGatewayEvent */
/** @typedef { import('aws-lambda').Context } Context */
import type { APIGatewayEvent, Context } from 'aws-lambda'

/**
 * The handler function is your code that processes http request events.
 * You can use return and throw to send a response or error, respectively.
 *
 * @param { APIGatewayEvent } event - an object which contains information from the invoker.
 * @param { Context } context - contains information about the invocation,
 * function, and execution environment.
 */
export const handler = async (event: APIGatewayEvent, context: Context) => {
  return {
    statusCode: 200,
    body: JSON.stringify({
      data: '${name} function',
    }),
  }
}
```

File: packages/cli/src/commands/generate/function/templates/function.ts.template [hunk 6]
```diff
@@ -0,0 +1,18 @@
+import type { APIGatewayEvent, Context } from 'aws-lambda'
+
+/**
+ * The handler function is your code that processes http request events.
+ * You can use return and throw to send a response or error, respectively.
+ * 
+ * @param { import('aws-lambda').APIGatewayEvent } event - an object which contains information from the invoker.
+ * @param { import('aws-lambda').Context } context - contains information about the invocation,
+ * function, and execution environment.
+ */
```
Comment (peterp) [inline packages/cli/src/commands/generate/function/templates/function.ts.template:10]:
```suggestion
 * @typedef { import('aws-lambda').APIGatewayEvent } APIGatewayEvent */
 * @typedef { import('aws-lambda').Context } Context */
 * @param { APIGatewayEvent } event - an object which contains information from the invoker.
 * @param { Context } context - contains information about the invocation,
 * function, and execution environment.
 */
```

Comment (peterp) [reply packages/cli/src/commands/generate/function/templates/function.ts.template:10]: It doesn't seem like I can submit this.

Comment (realStandal) [reply packages/cli/src/commands/generate/function/templates/function.ts.template:10]: Sorry! Totally assumed you could. I will right now.

**Review by peterp — APPROVED** [review]

**Comment by thedavidprice** [comment]
@realStandal This looks good to merge! I just tried to update from `main`, but it looks like I don't have push permissions. Could you update and then let me know? I'll get this in for our upcoming release.

Thanks again.

**Comment by realStandal** [comment]
> @realStandal This looks good to merge! I just tried to update from `main`, but it looks like I don't have push permissions. Could you update and then let me know? I'll get this in for our upcoming release.
> 
> Thanks again.

Of course! Does it not let y'all touch it because it's technically pushing to my repo, which GitHub is using to update this PR? So you'd need permissions on my repo? Just for my understanding.

**Comment by thedavidprice** [comment]
@realStandal When you create a PR, I believe there's a checkbox that's something like "allow redwoodjs/redwood maintainers to push to this branch" (or something like that). If that's not checked, then I _would_ need write permission for the fork.

I just missed the window between your lastest update and someone else merging new code. Could you try again? Or give me write permission? Thanks for the coordination! It's always the little things that get in the way, right? 😆

**Comment by thedavidprice** [comment]
@realStandal I see the push and am watching CI checks 🚀

**Comment by realStandal** [comment]
@thedavidprice Updating now, let's see if I can add you just in-case. Seems to be the case, I was thinking the same thing when I fixed the fixtures 10 times haha.

**Comment by realStandal** [comment]
oh you beat me to it!

**Comment by thedavidprice** [comment]
Got it in. Thanks again!

**Comment by realStandal** [comment]
Thank y'all for the opportunity! Hopefully I can find more to help out with

**Comment by thedavidprice** [comment]
You nailed a high priority via https://github.com/redwoodjs/redwood/issues/523 Anything else there you'd be interested in helping out with is a huge lift! Let us know if you have questions in the Issue, or open a new issue/draft PR as needed.

Thanks again!

**Pull request merged**